pkg/apis/networking/v1alpha1/clusteringress_types.go

```diff
@@ -0,0 +1,376 @@
+
+import (
+	"encoding/json"
+
+	"github.com/knative/pkg/apis"
+	"github.com/knative/pkg/apis/duck"
+	duckv1alpha1 "github.com/knative/pkg/apis/duck/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +genclient:nonNamespaced
+
+// ClusterIngress is a collection of rules that allow inbound connections to reach the
+// endpoints defined by a backend. An ClusterIngress can be configured to give services
+// externally-reachable urls, load balance traffic offer name based virtual hosting etc.
+//
+// This is heavily based on K8s Ingress https://godoc.org/k8s.io/api/extensions/v1beta1#Ingress
+// which some highlighted modifications.
+type ClusterIngress struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the desired state of the ClusterIngress.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Spec ClusterIngressSpec `json:"spec,omitempty"`
+
+	// Status is the current state of the ClusterIngress.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	// +optional
+	Status ClusterIngressStatus `json:"status,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterIngressList is a collection of ClusterIngress.
+type ClusterIngressList struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// Items is the list of ClusterIngress.
+	Items []ClusterIngress `json:"items"`
+}
+
+// ClusterIngressSpec describes the ClusterIngress the user wishes to exist.
+type ClusterIngressSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// TLS configuration. Currently the ClusterIngress only supports a single TLS
+	// port, 443. If multiple members of this list specify different hosts, they
+	// will be multiplexed on the same port according to the hostname specified
+	// through the SNI TLS extension, if the ingress controller fulfilling the
+	// ingress supports SNI.
+	// +optional
+	TLS []ClusterIngressTLS `json:"tls,omitempty"`
+
+	// A list of host rules used to configure the ClusterIngress.
+	// +optional
+	Rules []ClusterIngressRule `json:"rules,omitempty"`
```

Comment:
Thoughts on how we indicate whether the `ClusterIngress` should be externally exposed, and if so, on which network segment (i.e. RFC1918/on-prem/VPC vs internet)?